src/common/protocols/protocol_manager.h

```diff
@@ -0,0 +1,91 @@
+#ifndef GFS_COMMON_PROTOCOLS_PROTOCOL_MANAGER_H_
+#define GFS_COMMON_PROTOCOLS_PROTOCOL_MANAGER_H_
+
+#include <memory>
+
+#include "google/protobuf/stubs/status.h"
+#include "google/protobuf/stubs/statusor.h"
+#include "grpcpp/grpcpp.h"
+#include "src/protos/grpc/master_metadata_service.grpc.pb.h"
+
+namespace gfs {
+namespace protocols {
```

Comment:
Let's consider using "service"? And move the service client implementation to the same directory as service impl. Makes it easy to find.

Reply:
Renamed to `service`; however, will not move the file. 

It's not a matter of together; they are very different things, regardless of similar naming.

A client interface is a general common protocol manager client that can be used by anyone to talk to a gRPC "channel"; the gRPC classes are just built-in structures for serializing requests and deserializing replies, and client context, gRPC retries, etc. They do not even care about how the server handles it or implements it.

In the other hand, the `implementation` file is specifically for master to implement the handler. It doesn't **need** to be using gRPC service stubs (when can directly parse TCP messages sent to the port), but it's built-in for convenience. 

Thus, even though I started with the same idea as you, I think it makes more sense to have all the "client" in "common/protocols" for generic reuse communication purposes, while specific implementations reside in their corresponding folders.

Plus, it feel it makes a lot more sense for a library that can be used by both a client and chunk server to talk the master, to be in `common` rather than `src/server/master_server` as the latter should be specific to a master server

Reply:
Just seeing this. Don't agree with this and using the name "protocols" here definitely doesn't make more sense because it isn't self explanatory. I don't mind having it here but with a better name. I always push for names to be self explanatory. Just looking at a directory name, I should be able to guess what's in there. Don't want to delay this PR so it's approved.

Reply:
Propose a directory name then? Protocol client? Grpc client? 

Service client is kinda confusing tho, but I’m also ok with it

Reply:
renamed to `protocol_client` for directory, as this is not just for services alone but protocols client interface for all.